Source/Core/Core/PowerPC/JitArm64/JitArm64_RegCache.cpp

```diff
@@ -744,7 +744,7 @@ BitSet32 Arm64FPRCache::GetCallerSavedUsed() const
   for (const auto& it : m_host_registers)
   {
     if (it.IsLocked())
-      registers[it.GetReg() - Q0] = true;
+      registers[GetUsedRegisterIndex(it.GetReg())] = true;
```

Comment:
I believe the values in m_host_registers will all be Q registers which have values [192, 223].  Values in this range are mapped to [0, 31] both before and after this change.

Reply:
Yes, this change is correct. The function name `GetUsedRegisterIndex` sounds a bit unwieldy, though... Now that I think about it, is there actually any reason why `DecodeReg` shouldn't return an `int`?

Reply:
There are a number of places where DecodeReg is called and the return value is passed into another function that expects an ARM64Reg, so I wouldn't want to get rid of it.

On the other hand, there are a number of places where the return value is used in bitwise operations which effectively treat it as an int.

I'm not sure if this is best done in this PR or moved to another one, but we may want to create another function and have one return an Arm64Reg and the other an int.  I don't know what we would name them; DecodeReg already sounds a bit opaque to me even though I know what it does, so that might be an opportunity to give it a clearer name, but perhaps that name already is clear to people like you who are familiar with ARM assembly.

Reply:
> There are a number of places where DecodeReg is called and the return value is passed into another function that expects an ARM64Reg, so I wouldn't want to get rid of it.

After a quick look (I may have missed some instances), the functions I could find that were being called with the return value of `DecodeReg` were `UnlockRegister` (from JitArm64_RegCache) and `MOV`/`SXTW` (from JitArm64_Integer.cpp). In both of these cases, the `DecodeReg` call doesn't make sense and should just be removed as far as I can tell. (Things end up working anyway because the register in question is always a 32-bit GPR.)

Reply:
I started converting DecodeReg to return an int but found the change ended up being much more extensive than anticipated.

JitArm64::JumpIfCRFieldBit: calls DecodeReg and passes the result to either CBZ or CBNZ, which each insert the reg into a FixupBranch which is returned up to the bcx and bclrx functions, which pass the FixupBranch to SetJumpTarget, where the reg value is decoded a couple times and also passed to Is64Bit.

JitArm64::mcrxr: 
	calls LDRB, which calls one of two overloads of EncodeLoadStoreIndexedInst, which each call Is64Bit, IsVector, and DecodeReg. 
	calls ADD, which calls EncodeArithmeticInst, which calls Is64Bit and DecodeReg
	Initializes ArithOption, which initializes a member variable with it.

There are similar chains for crXXX, mfcr, mtcrf, cmpl, and cmpli in JitArm64, among others I haven't taken the time to investigate.

It would be possible to make all those changes, but it starts to approach the point of getting rid of ARM64Reg entirely.

Reply:
At least for `JumpIfCRFieldBit` and `mcrxr`, the appropriate thing to do would be to create a function `EncodeRegTo32` which works identically to how `DecodeReg` currently works. (Decoding a register and converting a register to its 32-bit form are two rather distinct operations semantically, but because the enum happens to start with `W0`, `reg & 0x1F` can be used to implement both.)

Also, a correction: In my previous comment, I didn't take into account that CR registers normally are mapped to 64-bit GPRs. Instead of removing the usages of `DecodeReg` that I mentioned in that comment, please replace them with `EncodeRegTo32`.

Reply:
Or in other words: Let's have both `DecodeReg` and `EncodeRegTo32`, with the former returning `int` and the latter returning `ARM64Reg`.